examples/rust/udp-push-pop.rs

```diff
@@ -95,173 +100,175 @@ fn freesga(libos: &mut LibOS, sga: demi_sgarray_t) {
+            Ok(sockqd) => sockqd,
+            Err(e) => anyhow::bail!("failed to create socket: {:?}", e),
         };
 
-        // Send data.
-        let qt: QToken = match libos.pushto(sockqd, &sga, remote) {
-            Ok(qt) => qt,
-            Err(e) => {
-                // If error, close socket and free scatter-gather array.
-                freesga(&mut libos, sga);
-                close(&mut libos, sockqd);
-                anyhow::bail!("push failed: {:?}", e)
-            },
-        };
-        match libos.wait(qt, None) {
-            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_PUSH => (),
-            Err(e) => {
-                // If error, close socket and free scatter-gather array.
-                freesga(&mut libos, sga);
-                close(&mut libos, sockqd);
-                anyhow::bail!("operation failed: {:?}", e)
-            },
-            _ => {
-                // If error, close socket and free scatter-gather array.
-                freesga(&mut libos, sga);
-                close(&mut libos, sockqd);
-                anyhow::bail!("unexpected result")
-            },
+        return Ok(Self {
+            libos,
+            sockqd,
+            sga: None,
+        });
+    }
+
+    fn run(
+        &mut self,
+        local: SocketAddrV4,
+        remote: SocketAddrV4,
+        fill_char: u8,
+        buffer_size: usize,
+        nsends: usize,
+    ) -> Result<()> {
+        match self.libos.bind(self.sockqd, local) {
+            Ok(()) => (),
+            Err(e) => anyhow::bail!("bind failed: {:?}", e),
         };
-        if let Err(e) = libos.sgafree(sga) {
-            // If error, close socket.
-            close(&mut libos, sockqd);
-            anyhow::bail!("failed to release scatter-gather array: {:?}", e)
+
+        // Issue n sends.
+        for i in 0..nsends {
+            self.sga = match mksga(&mut self.libos, buffer_size, fill_char) {
+                Ok(sga) => Some(sga),
+                Err(e) => anyhow::bail!("failed to allocate scatter-gather array: {:?}", e),
+            };
+
+            // Send data.
+            let qt: QToken = match self
+                .libos
+                .pushto(self.sockqd, &self.sga.expect("should be a valid sgarray"), remote)
+            {
+                Ok(qt) => qt,
+                Err(e) => anyhow::bail!("push failed: {:?}", e),
+            };
+
+            match self.libos.wait(qt, None) {
+                Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_PUSH => (),
+                Err(e) => anyhow::bail!("operation failed: {:?}", e),
+                _ => anyhow::bail!("unexpected result"),
+            };
+
+            match self.libos.sgafree(self.sga.expect("should be a valid sgarray")) {
+                Ok(_) => self.sga = None,
+                Err(e) => anyhow::bail!("failed to release scatter-gather array: {:?}", e),
+            }
+
+            println!("push ({:?})", i);
```

Comment:
You may avoid these calls to `except()`, but enclosing this in a `if let Some() { .. }` statement.